Static Viz and Exports/Downloads Respect Dashcard Viz Settings

adam-james-v opened the pull request:
WIP

Fixes: #45061

Downloads, subscriptions, and subscription attachments don't necessarily respect the viz-settings applied to a Dashboard Card.

This PR fixes discrepancies and adds test utils+tests to improve coverage in this area.

In this screenshot
TAXTAXTAX is a custom column name coming from the card.
CASH MONEY is a custom column name coming from the Dashcard.
Both get merged in correctly when downloading:
![image](https://github.com/metabase/metabase/assets/21064735/e8079aea-5286-4f4c-8ae5-340c0e71239d)

A subscription attachment also works.

adam-james-v committed:
start adding some util fns to download-exports-test ns

adam-james-v committed:
wip improving test util fns

adam-james-v committed:
add subscription attachment util fn

adam-james-v committed:
add tests to catch download/export differences

adam-james-v committed:
viz-settings from card and dashcard  merged properly

adam-james-v committed:
revert these changes which probably aren't needed

adam-james-v committed:
split test: downloads can be parallel, email attachments cannot

adam-james-v committed:
Merge branch 'master' into bugfix-45061-static-viz-and-exports-respect-dashcard-viz-settings

adam-james-v committed:
Merge branch 'master' into bugfix-45061-static-viz-and-exports-respect-dashcard-viz-settings

adam-james-v committed:
normalize viz-settings no matter their origin

piranha commented on test/metabase/api/downloads_exports_test.clj:
```diff
@@ -427,3 +539,69 @@
+            (is (= {:card-download     [card-header]
+                    :dashcard-download [dashcard-header]}
+                   {:card-download     (take 1 card-result)
+                    :dashcard-download (take 1 dashcard-result)}))))))))
```
`first`? 😁

adam-james-v replied:
😮‍💨 yea, of course haha

piranha commented on src/metabase/query_processor/middleware/visualization_settings.clj:
```diff
@@ -44,10 +44,11 @@
-      (when-let [card-id (-> query :info :card-id)]
-        (mb.viz/db->norm (:visualization-settings (lib.metadata.protocols/card (qp.store/metadata-provider) card-id))))))
+  (mb.viz/db->norm
+   (or (let [viz (-> query :viz-settings)]
+         (when (seq viz) viz))
```
it's not your code, but it could say `(-> query :viz-settings not-empty)` to be a bit easier to read.

adam-james-v replied:
Hmm.. I think this is actually to cover the case where :viz-settings IS there, but it's just an empty map, hence the `(seq viz)` check.

piranha replied:
`not-empty` does exactly that :) `(not-empty {}) ; => nil`

sansolovyov reviewed (approved):
Logic is cool, had a few non-blocking comments on code style in tests.

adam-james-v committed:
Update test/metabase/api/downloads_exports_test.clj

Co-authored-by: Alexander Solovyov <alexander@solovyov.net>

adam-james-v committed:
Incorporate some review feedback

adam-james-v committed:
rendered table cards also respect column names

adam-james-v committed:
Clean up table render tests and remove unhelpful render test utils

The big removal of code in the render test utils namespace is a clean up task that could've happened sooner, but this
table test adjustment is the perfect opportunity to clean up.

I had made the test utils before I truly knew how best to use `mt/with-temp`, so it had a lot of poorly implemented
'generators' that were doing a poor job of creating mock data for the rendering functions anyway.

So, reworking a few table tests here to just use a temp card is way more inline with the rest of the codebase, and
actually improves confidence that the tests are testing on well-formed data to begin with.
